Xplat/src/main/java/vazkii/botania/common/helper/EthicalTntHelper.java

```diff
@@ -0,0 +1,129 @@
+ */
+public class EthicalTntHelper {
+	/**
+	 * In case the various loaded dimensions' tick loops are run in their own threads, tracking data is kept
+	 * thread-local.
+	 */
+	private static final ThreadLocal<EthicalTntHelper> tracker = ThreadLocal.withInitial(EthicalTntHelper::new);
+
+	/**
+	 * While the moveBlocks method should not be reentrant, we can't know what crazy ideas some mixin authors might
+	 * have. To be safe, we account for reentrancy and hope that we don't miss any method exits. We also assume nobody
+	 * is crazy enough to spread the blocks-start-to-move logic across multiple threads.
+	 */
+	private final AtomicInteger trackTntEntities = new AtomicInteger();
+
+	/**
+	 * To reduce the risk of memory leaks, TNT entities are stored using their ID and a weak reference to their Level.
+	 */
+	private final WeakHashMap<Level, IntOpenHashSet> trackedTntEntities = new WeakHashMap<>();
+
+	/**
+	 * A force relay, force lens mana burst or vanilla piston is about to start moving blocks.
+	 */
+	public static void startTrackingTntEntities() {
+		tracker.get().startTracking();
+	}
+
+	/**
+	 * A TNT entity just spawned. Check it for potentially unethical spawning methods.
+	 *
+	 * @param entity The TNT entity.
+	 */
+	public static void addTrackedTntEntity(PrimedTnt entity) {
+		tracker.get().addTrackedEntity(entity);
+	}
+
+	/**
+	 * A force relay, force lens mana burst,or vanilla piston finished converting all blocks into moving block
+	 * entities.
+	 */
+	public static void endTrackingTntEntitiesAndCheck() {
+		tracker.get().endTracking();
+	}
+
+	private void startTracking() {
+		trackTntEntities.incrementAndGet();
+	}
+
+	private void addTrackedEntity(PrimedTnt entity) {
+		if (trackTntEntities.get() > 0) {
+			trackedTntEntities.computeIfAbsent(entity.getLevel(), lvl -> new IntOpenHashSet()).add(entity.getId());
+		}
+	}
+
+	private void endTracking() {
+		if (trackTntEntities.decrementAndGet() == 0) {
+			for (final var entry : trackedTntEntities.entrySet()) {
+				final var level = entry.getKey();
+				final var trackedEntities = entry.getValue();
+				if (trackedEntities != null) {
+					for (final var tntId : trackedEntities) {
+						final var entity = level.getEntity(tntId);
+						if (entity instanceof PrimedTnt tnt) {
+							checkUnethical(tnt);
+						}
+					}
+					trackedEntities.clear();
+				}
+			}
+		}
+	}
+
+	private static void checkUnethical(PrimedTnt entity) {
+		BlockPos center = entity.blockPosition();
+		if (!entity.getLevel().isLoaded(center)) {
+			return;
+		}
+
+		BlockPos.MutableBlockPos blockPos = new BlockPos.MutableBlockPos();
+		for (final var dir : Direction.values()) {
+			blockPos.setWithOffset(center, dir);
+			if (!entity.getLevel().isLoaded(blockPos)) {
+				continue;
+			}
+
+			final var blockState = entity.getLevel().getBlockState(blockPos);
+			if (blockState.getBlock() == Blocks.MOVING_PISTON && blockState.hasBlockEntity()) {
```

Comment:
.is() is the convention over ==, and the hasBlockEntity check will always be true

```suggestion
			if (blockState.is(Blocks.MOVING_PISTON)) {
```